cpp/src/arrow/compute/logical_type.h

```diff
@@ -0,0 +1,309 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Metadata objects for creating well-typed expressions. These are distinct
+// from (and higher level than) arrow::DataType as some type parameters (like
+// decimal scale and precision) may not be known at expression build time, and
+// these are resolved later on evaluation
+
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class DataType;
+class Status;
+
+namespace compute {
+
+class Expr;
+
+/// \brief An object that represents either a single concrete value type or a
+/// group of related types, to help with expression type validation and other
+/// purposes
+class ARROW_EXPORT LogicalType {
```

Comment:
Could we rename this TypeConstraint? That seems more descriptive of the intent.

Just for purposes of discussion: in the extreme case an instance would represent an arbitrary type requirement like:

```c++
auto signed_ints = type::anyOf(type::int8(), type::int16(), type::int32(), type::int64());
ASSERT_TRUE(signed_ints->SatisfiedBy(*scalar::int32(dummy)));
ASSERT_FALSE(signed_ints->SatisfiedBy(*array::binary(dummy)));

auto timestamped_struct = type::struct_({type::field("timestamp", type::timestamp())});
```

Even if we never need something as complex as that, I think the renaming will defuse a lot of user confusion along the lines of "well why isn't it just a DataType?"

Reply:
> Could we rename this TypeConstraint? 

Can we debate about names after we write some more code? Until we have more "real code" I think naming debates are not productive

Reply:
I think LogicalType is fine here, I think that constraints is not going to be the only thing where this is used.